#33: Read CallActivities from BPMN and integrate them in default.ftl template.

ahsNT opened the pull request.

ykstr commented on src/main/kotlin/info/novatec/cbdg/models/BpmnObject.kt:
```diff
@@ -1,19 +1,23 @@
+    val callActivityList: MutableList<CallActivity>
 ) {
     override fun toString(): String {
         return "${javaClass.name}[id=$id, name=$name, version=$version, documentation=$documentation, image=$image]"
```
Addition of callActivityList to toString?

ykstr replied:
Line 20: Avoid mutability whenever possible (Use List instead of MutableList unless you know you'll need to modify that list after its initial creation without re-creating it)

ykstr commented on src/main/kotlin/info/novatec/cbdg/models/elements/CallActivity.kt:
```diff
@@ -0,0 +1,4 @@
+package info.novatec.cbdg.models.elements
+
```
KDoc: What is a CallActivity?

ykstr commented on src/main/kotlin/info/novatec/docu/parser/element/CallActivityParser.kt:
```diff
@@ -0,0 +1,51 @@
+    fun parseToCallActivityList(process: Process): MutableList<CallActivity> {
+        val callActivityFlowElementList = process.flowElements.filter { it.elementType.typeName.equals(BPMN_TAG_CALL_ACTIVITY) }
+        val callActivityDocumentationList: MutableList<String> = mutableListOf()
+        callActivityFlowElementList.forEach { callActivityDocumentationList.add(it.documentations.stream().findFirst().orElse(null)?.textContent.toString()) }
```
use map instead of forEach to directly generate a non-mutable list

```
callActivityFlowElementList
   .map {
        CallActivity(
                        it.id, // Id of CallActivity
                        it.name, // Name of CallActivity
                        process.camundaVersionTag, // Version Tag of Process
                        it.documentations.findFirst().map{doc -> doc.textContent.toString()}.orElse(null), // Element Documentation of CallActivity
                        it.first.domElement.getAttribute(BPMN_ATTRIBUTE_CALLED_ELEMENT) // Called Element of CallAcitivity
                    )}

ykstr commented on src/main/resources/templates/process.ftl:
```diff
@@ -9,21 +9,25 @@
 <p>${bpmn.image}</p>
+<img src="../../${bpmn.image}">
 
+<li>CallActivities:</li>
```
`<li>`?

ahsNT commented:
Hi @ykstr ,
the PR is now updated and rebased.
Could you also have a quick look on the PR https://github.com/NovatecConsulting/cbdg-maven-test/pull/7 in test project? I've only changed the test bpmn file and it's the same change as in this PR.
Thanks! 🙂

ykstr commented on src/main/kotlin/info/novatec/docu/parser/element/CallActivityParser.kt:
```diff
@@ -0,0 +1,31 @@
+     * comprised in one object.
+     */
+    fun parseToCallActivityList(process: Process): List<CallActivity> {
+        val callActivityFlowElementList = process.flowElements.filter { it.elementType.typeName.equals(BPMN_TAG_CALL_ACTIVITY) }
```
Kotlin allows [operator overloading](https://kotlinlang.org/docs/operator-overloading.html), so you can use == instead of equals() to compare objects because the == is overloaded with a call to equals.
This provides the additional benefit of null safety since according to the documentation == is always true for null == null.

Doesnt necessarily need to be changed, but its nice to know.
